module/execution/scripts.go

```diff
@@ -0,0 +1,119 @@
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+// - ErrIndexBoundary if the height is out of indexed height boundary
+type RegistersAtHeight func(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error)
+
+type Scripts struct {
+	executor          *query.QueryExecutor
+	headers           storage.Headers
+	registersAtHeight RegistersAtHeight
+}
+
+func NewScripts(
+	log zerolog.Logger,
+	metrics *metrics.ExecutionCollector,
+	chainID flow.ChainID,
+	entropy query.EntropyProviderPerBlock,
+	header storage.Headers,
+	registersAtHeight RegistersAtHeight,
+) (*Scripts, error) {
+	vm := fvm.NewVirtualMachine()
+
+	options := computation.DefaultFVMOptions(chainID, false, false)
+	vmCtx := fvm.NewContext(options...)
+
+	derivedChainData, err := derived.NewDerivedChainData(derived.DefaultDerivedDataCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	queryExecutor := query.NewQueryExecutor(
+		query.NewDefaultConfig(),
+		log,
+		metrics,
+		vm,
+		vmCtx,
+		derivedChainData,
+		entropy,
+	)
+
+	return &Scripts{
+		executor:          queryExecutor,
+		headers:           header,
+		registersAtHeight: registersAtHeight,
+	}, nil
+}
+
+// ExecuteAtBlockHeight executes provided script against the block height.
+// A result value is returned encoded as byte array. An error will be returned if script
+// doesn't successfully execute.
+// Expected errors:
+// - Storage.NotFound if block or register value at height was not found.
+func (s *Scripts) ExecuteAtBlockHeight(
+	ctx context.Context,
+	script []byte,
+	arguments [][]byte,
+	height uint64) ([]byte, error) {
+
+	snap, header, err := s.snapshotWithBlock(height)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.executor.ExecuteScript(ctx, script, arguments, header, snap)
+}
+
+// GetAccount returns a Flow account by the provided address and block height.
+// Expected errors:
+// - Storage.NotFound if block or register value at height was not found.
+func (s *Scripts) GetAccount(ctx context.Context, address flow.Address, height uint64) (*flow.Account, error) {
+	snap, header, err := s.snapshotWithBlock(height)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.executor.GetAccount(ctx, address, header, snap)
+}
+
+// snapshotWithBlock is a common function for executing scripts and get account functionality.
+// It creates a storage snapshot that is needed by the FVM to execute scripts.
+func (s *Scripts) snapshotWithBlock(height uint64) (snapshot.StorageSnapshot, *flow.Header, error) {
+	header, err := s.headers.ByHeight(height)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	storageSnapshot := snapshot.NewReadFuncStorageSnapshot(func(ID flow.RegisterID) (flow.RegisterValue, error) {
+		values, err := s.registersAtHeight(flow.RegisterIDs{ID}, height)
```

Comment:
we only get one register at a time, do we need `registersAtHeight` or just `registerAtHeight`?

Reply:
I just thought I would make the function compatible with the indexer existing function, I can make an adapter and make this function only take a single register. I personally don't mind either way. One adds a bit more code the other is not as perfect since we always just need one like you said.